ios/app/JitsiMeetBroadcast Extension/SampleUploader.m

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright @ 2018-present 8x8, Inc.
+ * Copyright @ 2017-2018 Atlassian Pty Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <MessageUI/MessageUI.h>
+#import <ReplayKit/ReplayKit.h>
+
+#import "SampleUploader.h"
+#import "SocketConnection.h"
+
+static const NSInteger kBufferMaxLenght = 1000 * 1024;
+
+@interface SampleUploader ()
+
+@property (nonatomic, assign) BOOL isReady;
+
+@property (nonatomic, strong) dispatch_queue_t serialQueue;
+@property (nonatomic, strong) SocketConnection *connection;
+@property (nonatomic, strong) CIContext *imageContext;
+
+@property (nonatomic, strong) NSData *dataToSend;
+@property (nonatomic, assign) NSUInteger byteIndex;
+
+@end
+
+@implementation SampleUploader
+
+- (instancetype)initWithConnection:(SocketConnection *)connection {
+  self = [super init];
+  if (self) {
+    self.serialQueue = dispatch_queue_create("org.jitsi.meet.broadcast.sampleUploader", DISPATCH_QUEUE_SERIAL);
+    
+    self.connection = connection;
+    [self setupConnection];
+    
+    self.imageContext = [[CIContext alloc] initWithOptions:nil];
+    self.isReady = false;
+  }
+  
+  return self;
+}
+
+- (void)sendSample:(CMSampleBufferRef)sampleBuffer {
+  self.isReady = false;
+  
+  [self prepareSample:sampleBuffer];
+  
+  dispatch_async(self.serialQueue, ^{
+    [self sendData];
+  });
+}
+
+// MARK: Private Methods
+
+- (void)setupConnection {
+  __weak __typeof(self) weakSelf = self;
+  self.connection.didOpen = ^{
+    weakSelf.isReady = true;
+  };
+  self.connection.streamHasSpaceAvailable = ^{
+    dispatch_async(weakSelf.serialQueue, ^{
+      weakSelf.isReady = ![weakSelf sendData];
+    });
+  };
+}
+
+- (void)prepareSample:(CMSampleBufferRef)sampleBuffer {
```

Comment:
Document what this does please.

Reply:
done. Also, I've slightly updated the logic so this function only does the image processing and wrapping into a CFHTTPMessageRef. Storing the data & reseting the state part is moved to `sendSample: `